pb-rs/src/types.rs

```diff
@@ -847,19 +847,31 @@ impl Message {
             writeln!(w, "")?;
             writeln!(w, "pub mod mod_{} {{", self.name)?;
             writeln!(w, "")?;
-            if self
-                .messages
-                .iter()
-                .any(|m| m.all_fields().any(|f| f.typ.has_cow()))
-            {
-                writeln!(w, "use std::borrow::Cow;")?;
+            if config.nostd {
+                writeln!(w, "extern crate alloc;")?;
+                writeln!(w, "use alloc::vec::Vec;")?;
+            }
+            if self.messages.iter().any(|m| {
+                m.all_fields()
+                    .any(|f| (f.typ.has_cow() || (f.packed() && f.typ.is_fixed_size())))
+            }) {
+                if config.nostd {
+                    writeln!(w, "use alloc::borrow::Cow;")?;
+                } else {
+                    writeln!(w, "use std::borrow::Cow;")?;
+                }
             }
             if self
                 .messages
                 .iter()
                 .any(|m| m.all_fields().any(|f| f.typ.is_map()))
             {
-                writeln!(w, "use std::collections::HashMap;")?;
+                if config.nostd {
+                    writeln!(w, "use alloc::collections::BTreeMap;")?;
+                    writeln!(w, "pub type HashMap<K, V> = BTreeMap<K, V>;")?;
```

Comment:
It seems good to have client code not care about which mode the proto is compiled in, in the default case.  Why do you prefer to have separate APIs?

Reply:
I don't like misleading APIs. `HashMap` conveys a O(1) and a Hash key instead of a O(log(n)) and a Ord key.

Reply:
I kinda like the suggestion of depending on hashbrown directly. Would adding that dependency in a no_std env be acceptable?

Reply:
i actually dont like this at all. in general in no_std you dont want too many dependencies, because the status of your code being 'no_std' depends on all the dependencies (including transitive ones) being no_std, which can get annoying. i see the utility of having support for all protobuf types, but i still think that you should feature-gate this, and have the dependency as optional for this feature (like features = ["hashmap"], etc.)

Reply:
A feature gate makes sense, though I think we can probably just do a build flag instead since this is just for generated code. `pb-rs --hashmap` or something like that. Would you want it to error if the proto has a map type but `--hashmap` isn't enabled?

Reply:
Also, I assume `--hashmap` would use `hashbrown` or something similar if enabled?

Hmmm. We would need a feature-gate + the `pb-rs` flag I think. Feature-gate for quick-protobuf and flag for `pb-rs`....

Reply:
yeah exactly, i agree on the flag + feature. as for errors, it would make most sense to error in case of mismatch. the more interesting way to deal with this (IMHO) is actually given in the protobuf api itself. see https://developers.google.com/protocol-buffers/docs/proto#maps, under "Backwards compatibility". especially the requirement that "Any protocol buffers implementation that supports maps must both produce and accept data that can be accepted by the above definition."  Which technically means we can still be wire-compatible even without depending on anything. this is arguably more work to implement, and is probably outside the scope of this PR, but something to think about.

Reply:
True. I hadn't thought about the maps-are-really-just-a-repeated-field aspect. I like that as the default. I suppose we'd just create a synthetic nested message for the map field in the generated code.